### PR TITLE
add GitHub Action to build Mixxx on macOS

### PR DESCRIPTION
**Pull request by Be-ing**

### CODE REVIEW

File: .github/workflows/mixxx.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+name: mixxx
```
Comment (Holzhaus) [inline .github/workflows/mixxx.yml:1]:
```suggestion
name: Build
```

?

File: .github/workflows/mixxx.yml [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+on: [pull_request, push]
+
+jobs:
+  build-mac:
+    runs-on: macos-10.14
```
Comment (Holzhaus) [inline .github/workflows/mixxx.yml:7]:
```suggestion
    name: clang on macOS
    runs-on: macos-10.14
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Thanks. If this works I can make a follow-up PR to clean this up a little.

File: .github/workflows/mixxx.yml (hunk 2 above)
Comment (Be-ing) [reply .github/workflows/mixxx.yml:7]: why?

File: .github/workflows/build-mixxx.yml [hunk 3]
```diff
@@ -59,7 +59,7 @@ jobs:
         cmake -L $CMAKEFLAGS $CMAKEFLAGS_EXTRA -DMACOS_BUNDLE=ON -DMACOS_MIN_VERSION=10.12 -DCMAKE_PREFIX_PATH=${MIXXX_ENVPATH} -DQt5_DIR=${QT_DIR} ..
         cmake --build .
         cmake --build . --target benchmark
-        ctest --timeout 45
+        #ctest --timeout 45
```
Comment (uklotzde) [inline .github/workflows/build-mixxx.yml:62]:
```
# TODO: Fix disabled test(s)
ctest --timeout 45 --exclude-regex DirectoryDAOTest.relocateDirectory
```

**Comment by Be-ing** [comment]
I added support for code signing, but I cannot test it until @rryan adds the MACOS_CODESIGN_OPENSSL_PASSWORD and MACOS_CODESIGN_CERTIFICATE_PASSWORD secrets.

**Comment by uklotzde** [comment]
Are the secrets for the code signing step the only thing that is missing now? Should we merge anyway? The worst thing that could happen is that the build breaks after those secrets have been added. But in this case we could easily disable the whole code signing step temporarily.

I suggest to abandon the Travis CI builds as soon as this PR has been merged.

**Comment by Be-ing** [comment]
> Are the secrets for the code signing step the only thing that is missing now?

Yes.

> Should we merge anyway?

I was hoping @Holzhaus could take a look before that since this is the first GitHub Action file I've written, but changes can be made after merging too.

> I suggest to abandon the Travis CI builds as soon as this PR has been merged.

I agree.

**Comment by Be-ing** [comment]
Deployment to downloads.mixxx.org is still missing too but that also needs the secret credentials added. In the meantime at least we have the artifacts hosted with GitHub Actions.

**Comment by uklotzde** [comment]
@Holzhaus You have requested changes. Resolved or not?

File: .github/workflows/build-mixxx.yml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+      MIXXX_ENVNAME: 2.3-1db6c06-sdk10.15-macosminimum10.12-x86_64
+      # macOS 10.12 is the minimum requirement for Qt 5.12.
+      # Check the minimum requirement when updating to a build environment
+      # with an updated Qt version.
+      MACOSX_DEPLOYMENT_TARGET: 10.12
+      CMAKEFLAGS_EXTRA: -DCOREAUDIO=ON -DHSS1394=ON -DMACOS_BUNDLE=ON
+      GTEST_COLOR: 1
+      CTEST_OUTPUT_ON_FAILURE: 1
+      # Render analyzer waveform tests to an offscreen buffer
+      QT_QPA_PLATFORM: offscreen
```
Comment (Holzhaus) [inline .github/workflows/build-mixxx.yml:25]: This can be moved to a separate ctest step

File: .github/workflows/build-mixxx.yml [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  CMAKEFLAGS: -DCMAKE_BUILD_TYPE=Release -DBATTERY=ON -DBROADCAST=ON -DBULK=ON -DDEBUG_ASSERTIONS_FATAL=ON -DHID=ON -DLILV=ON -DOPUS=ON -DQTKEYCHAIN=ON -DVINYLCONTROL=ON
+  # GitHub Actions automatically zstd compresses caches
+  CCACHE_NOCOMPRESS: true
+
+jobs:
+  build-mac:
+    name: macOS
+    runs-on: macos-10.15
+    env:
+      MIXXX_ENVNAME: 2.3-1db6c06-sdk10.15-macosminimum10.12-x86_64
```
Comment (Holzhaus) [inline .github/workflows/build-mixxx.yml:16]: TODO parse this from the repo file

File: .github/workflows/build-mixxx.yml [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        cd cmake_build
+        cmake -L $CMAKEFLAGS $CMAKEFLAGS_EXTRA -DCMAKE_PREFIX_PATH=${MIXXX_ENVPATH} -DQt5_DIR=${QT_DIR} ..
+        cmake --build . --parallel $(sysctl -n hw.physicalcpu)
+        cmake --build . --target benchmark
+        #TODO: fix disabled test
+        ctest --timeout 45 --exclude-regex DirectoryDAOTest.relocateDirectory
+        ccache -s
+    - name: package app bundle
+      run: |
+        cd /Users/runner/work/mixxx/mixxx/cmake_build
```
Comment (Holzhaus) [inline .github/workflows/build-mixxx.yml:99]: You use use `working_directory: cmake_build` for that

File: .github/workflows/build-mixxx.yml [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        fi
+
+        cd /Users/runner/work/mixxx/mixxx
+        mkdir cmake_build
+        cd cmake_build
+        cmake -L $CMAKEFLAGS $CMAKEFLAGS_EXTRA -DCMAKE_PREFIX_PATH=${MIXXX_ENVPATH} -DQt5_DIR=${QT_DIR} ..
+        cmake --build . --parallel $(sysctl -n hw.physicalcpu)
+        cmake --build . --target benchmark
+        #TODO: fix disabled test
+        ctest --timeout 45 --exclude-regex DirectoryDAOTest.relocateDirectory
```
Comment (Holzhaus) [inline .github/workflows/build-mixxx.yml:95]: Split this into a separate test step? Makes it easier to see what exactly failed.

**Review by Holzhaus — COMMENTED** [review]
Thanks. Just a few comments before we can merge. I'll look into combining this with the Ubuntu build and check if we can use a build matrix to reduce duplication soon.

File: .github/workflows/build-mixxx.yml (hunk 4 above)
Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: Unfortunately it cannot. When I tried to do that, every test SIGABORTed.

File: .github/workflows/build-mixxx.yml (hunk 5 above)
Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: Why?

Comment (Holzhaus) [reply .github/workflows/build-mixxx.yml]: That's how we always did it until now (`build/osx/golden_environment`). It makes it  possible tp add helper scripts for each OS that fetch and set up the environment (CMAKE_PREFIX_PATH) for the current machine. That could make it easier for users to configure their mixxx build properly and we could reuse them here.

Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: I suppose it does make it easier to look back in the Git history when the build environment was updated. Changing that old file will break Travis builds (and Jenkins, but that's already disabled). Shall I go ahead and disable Travis in this PR?

File: .github/workflows/build-mixxx.yml (hunk 4 above)
Comment (Holzhaus) [reply .github/workflows/build-mixxx.yml]: So this does not work? https://github.com/mixxxdj/mixxx/blob/2.3/.github/workflows/build.yml#L95L101

Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: I'll give it another try, but when I tried before it just crashed.

Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: I tried again and it [failed](https://github.com/mixxxdj/mixxx/runs/1420695792?check_suite_focus=true#step:7:3748).

File: .github/workflows/build-mixxx.yml (hunk 5 above)
Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: I will hack around this by making a new file in the `cmake` directory instead of the old `build/osx`

Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: Nevermind. I do not think we should do this. The build environment name is [used for a cache key](https://github.com/mixxxdj/mixxx/pull/3305/files#diff-5c3fa597431eda03ac3339ae6bf7f05e1a50d6fc7333679ec38e21b337cb6721R128). I don't know how we could do that if it had to be read from a file.

File: .github/workflows/build-mixxx.yml (hunk 4 above)
Comment (Holzhaus) [reply .github/workflows/build-mixxx.yml]: Weird.

**Comment by Be-ing** [comment]
Ready for merge?

File: .github/workflows/build-mixxx.yml (hunk 5 above)
Comment (Holzhaus) [reply .github/workflows/build-mixxx.yml]: It's possible. Set `id: setup_buildenv` of the step and print this to the terminal during `run`. 

    ::set-output name=buildenv_name::my-custom-buildenv-name

In a later step, you can then do:

    key: buildenv-cache-${{ steps.setup_buildenv.outputs.buildenv_name }}

Comment (Holzhaus) [reply .github/workflows/build-mixxx.yml]: Anyway, no need to do it now, just leave a TODO comment.

Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: That seems overcomplicated for very little benefit.

**Comment by Be-ing** [comment]
> I'll look into combining this with the Ubuntu build and check if we can use a build matrix to reduce duplication soon.

I suspect that will create as much complexity as the redundancy it would remove.

File: .github/workflows/build-mixxx.yml (hunk 5 above)
Comment (Holzhaus) [reply .github/workflows/build-mixxx.yml]: Making the the dev setup easier for contributors is definitely worth 3 additional lines in a CI config file IMHO.

**Comment by Holzhaus** [comment]
> > I'll look into combining this with the Ubuntu build and check if we can use a build matrix to reduce duplication soon.
> 
> I suspect that will create as much complexity as the redundancy it would remove.

Possible, but let's try and decide then.

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -4,9 +4,18 @@ on:
...
   pull_request:
 
+env:
+  CMAKEFLAGS_COMMON: -DCMAKE_BUILD_TYPE=Release -DBATTERY=ON -DBROADCAST=ON -DBULK=ON -DDEBUG_ASSERTIONS_FATAL=ON -DHID=ON -DLILV=ON -DOPUS=ON -DQTKEYCHAIN=ON -DVINYLCONTROL=ON
+  GTEST_COLOR: 1
+  CTEST_OUTPUT_ON_FAILURE: 1
+  # Render analyzer waveform tests to an offscreen buffer
+  QT_QPA_PLATFORM: offscreen
+  # GitHub Actions automatically zstd compresses caches
+  CCACHE_NOCOMPRESS: true
```
Comment (Holzhaus) [inline .github/workflows/build.yml:14]: Is this a time-memory tradeoff? Does disabling compression really make a big difference in build time?

File: .github/workflows/build-mixxx.yml (hunk 5 above)
Comment (Be-ing) [reply .github/workflows/build-mixxx.yml]: Okay, done. I also got the macOS minimum version number from the build environment name so this CI script doesn't need to be updated when that changes.

**Comment by Be-ing** [comment]
Ready for merge?

**Review by Holzhaus — APPROVED** [review]

**Pull request merged**